File: hash_table.rb
class HashTable
  attr_accessor :table

  def initialize
    @table = Array.new(10) { |x| x = LinkedList.new }
  end

  # what is that key for?
  def put(_key, value)
    value = value.to_s
    index = value.size

    @table[index].insert(value)
  end

  def get(value)
    index = value.size

    @table[index].find(value)
  end

  def tprint 
    @table.each.with_index do |item, index|
      print "#{index} -> "
      item.print_list
      print "\n"
    end
  end
end

class LinkedList 
  attr_accessor :data, :next
  
  def initialize(data = nil)
    @data = data
    @next = nil
  end

  def print_list
    if @data
      print "#{@data}, "
    end

    if @next
      @next.print_list
    end
  end

  def insert(value)
    if @data.nil?
      @data, @next = value, LinkedList.new
    else
      return @next.insert(value)
    end

    return self
  end

  def find(value)
    if @data.nil?
      return "Nonexistant"
    else
      @data == value ? data : @next.find(value)
    end
  end
end

File: hash_table.go
package main

import (
    "fmt"
)

type Llist struct {
    next *Llist
    data string
}

func main() {
    array := make([]Llist, 10)

    // Setup
    insert(array, "Prestige")
    insert(array, "Go")
    insert(array, "Dont")
    insert(array, "Wont")
    insert(array, "Hash")
    insert(array, "Table")
    insert(array, "rm -rf *")
    insert(array, "Oh")

    // Prints the Linked List key value pairs
    print(array)
}

//func (l *Llist) find(data string) (*Llist, error) {
    //TODO
        //Traverse List until desired value is found
        //Return error if value does not exist
//}

// Prints a Linked List
func print(array []Llist) {
    for i := 0; i < len(array); i++ {
        fmt.Printf("%d -> ", i)
        PrintList(array[i])
        fmt.Println("")
    }
}

// Helper for print
func PrintList(llist Llist) {
    if len(llist.data) > 0 {
        fmt.Printf("%s, ", llist.data)
    }

    if llist.next != nil {
        PrintList(*llist.next)
    }
}

// Insert into an array of Linked Lists
func insert(array []Llist, data string) *Llist {
    // Simple hash index calc
    index := len(data)
    return array[index].insert(data)
}

// Recursive Insertion
func (l *Llist) insert(data string) *Llist {
    if len(l.data) < 1 {
        l.data = data
        l.next = &Llist{}
    } else {
        l = l.next.insert(data)
    }

    return l
}
